Modules/atexitmodule.c

```diff
@@ -99,7 +120,10 @@ void
 _PyAtExit_Fini(PyInterpreterState *interp)
 {
     struct atexit_state *state = &interp->atexit;
-    atexit_cleanup(state);
+    // Only one thread can call this, no need to lock it
+    _PyAtExit_ASSERT_UNLOCKED(state);
+
+    atexit_cleanup_locked(state);
```

Comment:
atexit_delete_cb_locked() does a unlock/re-lock dance. It seems unsafe to call this "locked" function without holding the lock. I suggest to add LOCK/UNLOCK around this call.